src/main/java/org/myrobotlab/service/Mqtt.java

```diff
@@ -1,439 +1,538 @@
+		// Called when a message arrives from the server that matches any
+		// subscription made by the client
+		String time = new Timestamp(System.currentTimeMillis()).toString();
+		// FIXME - new String won't be happy with true binary payloads ..
+		String payload = new String(message.getPayload());
+		String messageStr = "onMqttMsg Time: " + time + "\tTopic: " + topic + "\tMessage: " + payload + "\tQoS: "
+				+ message.getQos();
+		log.info(messageStr);
+
+		// serialization needs to be a logical layer - so we can change it to
+		// protobuf or native Java
+		try {
+			Message msg = (Message) codec.decode(payload, Message.class);
+
+			// COMMON GATEWAY REGISTERATION AND X-FORWARDED BEGIN --------------
+
+			// make a mrl key and protocol uri begin ------
+			// ADD TCP CLIENT BEGIN ! - probably should not be in RemoteAdapter - as this is
+			// a detail for tcp
+
+			// String clientKey = String.format("mqtt://%s:%d", this.url);
+			URI uri = new URI(url);
+			// HELP PROTOKEY VS MRL KEY ??
+			// TcpThread tcp = new TcpThread(myService, uri, clientSocket);
+			// tcpClientList.put(uri, tcp);
+			// myService.connections.put(uri, tcp.data);
+
+			// make a mrl key and protocol uri end -------
+
+			// registration request - security is applied here
+			// x-forwarded is
+			// msg.name == runtime;
+
+			// what if you don't want to do x-forwarding ???
+			inbox.add(msg);
+
+			// COMMON GATEWAY REGISTERATION AND X-FORWARDED END ----------------
+
+		} catch (Exception e) {
+			error("tried decoding [%] into a message", payload);
+		}
+
+		// are all necessary ?
+		invoke("publishMqttMsg", topic, message.getPayload());
+		invoke("publishMqttMsgByte", message.getPayload());
+		invoke("publishMqttMsgString", new String(message.getPayload()), topic);
+	}
+
+	@Override
+	public void onFailure(IMqttToken token, Throwable throwable) {
+		log.error("error - {} {}", tokenToString(token), throwable);
+	}
+
+	@Override
+	public void onSuccess(IMqttToken token) {
+		log.info("success - {} ", tokenToString(token));
+	}
+
+	// dangerous - as topic is a field value
+	public void publish(String msg) throws Throwable {
+		publish(topic, qos, msg);
+	}
+
+	public void publish(String topic, Integer qos, byte[] payload) throws MqttPersistenceException, MqttException {
+		MqttMessage message = null;
+
+		if (client == null || !client.isConnected()) {
+			connect(url);
+		}
+
+		// Send / publish a message to the server
+		// Get a token and setup an asynchronous listener on the token which
+		// will be notified once the message has been delivered
+		message = new MqttMessage(payload);
+		message.setQos(qos);
+
+		String time = new Timestamp(System.currentTimeMillis()).toString();
+		log.info("Publishing at: " + time + " to topic \"" + topic + "\" qos " + qos);
+
+		// FIXME - see if user context is like a backend id ...
+		client.publish(topic, message, "Pub sample context", this);
```

Comment:
`"Pub sample context"` seems like a weird hardcoded value.

Reply:
Aye .. its a "userContext" and it came from a sample, but I'm unclear at the moment of its purpose

Reply:
done